### PR TITLE
feat(cloud-provisioning): Adds function to provision certificate to existing machine identity

### PR DESCRIPTION
**Pull request by rvelaVenafi**
Refactors websocket functionality into a notificationServiceClient

### CODE REVIEW

File: pkg/domain/workflow.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+type WorkFlowResponseData struct {
+	Result       interface{} `json:"result"`
+	WorkflowId   string      `json:"workflowId"`
+	WorkflowName string      `json:"workflowName"`
+	WsClientId   string      `json:"wsClientId"`
+}
+
+type WorkflowResponse struct {
+	SpecVersion     string               `json:"specversion"`
+	Id              string               `json:"id"`
```
Comment (sbanic-venafi) [inline pkg/domain/workflow.go:12]: nit: You have no linter to complain about "ID" vs "Id" : -). Resolve this if you don't care.

Comment (rvelaVenafi) [reply pkg/domain/workflow.go:12]: Fixed

File: pkg/domain/workflow.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+	Source          string               `json:"source"`
+	Type            string               `json:"type"`
+	Subject         string               `json:"subject"`
+	DataContentType string               `json:"datacontenttype"`
+	Time            string               `json:"time"`
+	Data            WorkFlowResponseData `json:"data"`
+	EventKind       string               `json:"eventkind"`
+	EventResource   string               `json:"eventresource"`
+	Recipient       string               `json:"recipient"`
+	CorrelationId   string               `json:"correlationid"`
```
[the remark this answers is not in the document]
Comment (rvelaVenafi) [reply pkg/domain/workflow.go:22]: Fixed

File: pkg/venafi/cloud/cloudproviders.go [hunk 3]
```diff
@@ -229,6 +230,187 @@ func (c *Connector) getGraphqlHTTPClient() *http.Client {
...
+		if reqData.KeystoreID == nil {
+			if reqData.ProviderName == nil || reqData.KeystoreName == nil {
+				return nil, fmt.Errorf("any of keystore object, keystore ID or both Provider Name and Keystore Name must be provided for provisioning")
+			}
+		}
+
+		// Getting Keystore to find type
+		keystoreIDInput := util.StringPointerToString(reqData.KeystoreID)
+		keystoreNameInput := util.StringPointerToString(reqData.KeystoreName)
+		providerNameInput := util.StringPointerToString(reqData.ProviderName)
```
[the remark this answers is not in the document]
Comment (rvelaVenafi) [reply pkg/venafi/cloud/cloudproviders.go:293]: GetCloudKeystore returns an appropriate error. All attributes are optional.

File: pkg/venafi/cloud/cloudproviders.go [hunk 4]
```diff
@@ -229,6 +230,187 @@ func (c *Connector) getGraphqlHTTPClient() *http.Client {
...
+
+	if req.MachineIdentityID == nil {
+		return nil, fmt.Errorf("error trying to provision certificate to machine identity: machineIdentityID is nil")
+	}
+
+	machineIdentityID := *req.MachineIdentityID
+	certificateID := ""
+	timeout := util.DefaultTimeout * time.Second
+	if req.Timeout != 0 {
+		timeout = req.Timeout
```
[the remark this answers is not in the document]
Comment (rvelaVenafi) [reply pkg/venafi/cloud/cloudproviders.go:355]: `req.Timeout` is already a `time.Duration` variable. The user is responsible for passing the correct duration

File: pkg/venafi/cloud/cloudproviders.go [hunk 5]
```diff
@@ -229,6 +230,187 @@ func (c *Connector) getGraphqlHTTPClient() *http.Client {
...
+func (c *Connector) ProvisionCertificate(req *endpoint.ProvisioningRequest, options *endpoint.ProvisioningOptions) (provisioningMetadata endpoint.ProvisioningMetadata, err error) {
+	log.Printf("Starting Provisioning Flow")
+
+	if req == nil {
+		return nil, fmt.Errorf("missing Provisioning Request")
+	}
+
+	reqData := *req
+
+	if reqData.Timeout == 0 {
```
[the remark this answers is not in the document]
Comment (rvelaVenafi) [reply pkg/venafi/cloud/cloudproviders.go:242]: no need. `time.Duration`'s underlying primitive is int. The comparison only serves to find out if any timeout has been passed by the user

**Review by LaloHao — APPROVED** [review]

**Review by marcos-albornoz — APPROVED** [review]

**Review by EduardoVV — APPROVED** [review]

**Review by luispresuelVenafi — APPROVED** [review]

**Pull request merged**